packages/react/popper/src/Popper.tsx

```diff
@@ -105,71 +126,136 @@ const PopperContent = React.forwardRef<PopperContentElement, PopperContentProps>
       side = 'bottom',
-      sideOffset,
+      sideOffset = 0,
       align = 'center',
-      alignOffset,
-      collisionTolerance,
+      alignOffset = 0,
+      arrowPadding = 0,
+      collisionTolerance = 0,
       avoidCollisions = true,
       ...contentProps
     } = props;
 
     const context = usePopperContext(CONTENT_NAME, __scopePopper);
-    const [arrowOffset, setArrowOffset] = React.useState<number>();
-    const anchorRect = useRect(context.anchor);
+
     const [content, setContent] = React.useState<HTMLDivElement | null>(null);
-    const contentSize = useSize(content);
+    const composedRefs = useComposedRefs(forwardedRef, (node) => setContent(node));
+
     const [arrow, setArrow] = React.useState<HTMLSpanElement | null>(null);
     const arrowSize = useSize(arrow);
+    const arrowWidth = arrowSize?.width ?? 0;
+    const arrowHeight = arrowSize?.height ?? 0;
+
+    const desiredPlacement = (side + (align !== 'center' ? '-' + align : '')) as Placement;
+
+    const { reference, floating, strategy, x, y, placement, middlewareData, update } = useFloating({
+      // default to `fixed` strategy so users don't have to pick and we also avoid focus scroll issues
+      strategy: 'fixed',
+      placement: desiredPlacement,
+      whileElementsMounted: autoUpdate,
+      middleware: [
+        offset({ mainAxis: sideOffset + arrowHeight, alignmentAxis: alignOffset }),
+        avoidCollisions
+          ? shift({
+              mainAxis: true,
+              crossAxis: false,
+              padding: collisionTolerance,
+              limiter: limitShift(),
+            })
+          : undefined,
+        arrow ? floatingUIarrow({ element: arrow, padding: arrowPadding }) : undefined,
+        avoidCollisions ? flip({ padding: collisionTolerance }) : undefined,
+        transformOrigin({ arrowWidth, arrowHeight }),
+      ].filter(isDefined),
+    });
 
-    const composedRefs = useComposedRefs(forwardedRef, (node) => setContent(node));
+    // assign the reference dynamically once `Content` has mounted so we can collocate the logic
+    useLayoutEffect(() => {
+      reference(context.anchor);
+    }, [reference, context.anchor]);
 
-    const windowSize = useWindowSize();
-    const collisionBoundariesRect = windowSize
-      ? DOMRect.fromRect({ ...windowSize, x: 0, y: 0 })
-      : undefined;
-
-    const { popperStyles, arrowStyles, placedSide, placedAlign } = getPlacementData({
-      anchorRect,
-      popperSize: contentSize,
-      arrowSize,
-
-      // config
-      arrowOffset,
-      side,
-      sideOffset,
-      align,
-      alignOffset,
-      shouldAvoidCollisions: avoidCollisions,
-      collisionBoundariesRect,
-      collisionTolerance,
-    });
-    const isPlaced = placedSide !== undefined;
+    const isPlaced = x !== null && y !== null;
+    const [placedSide, placedAlign] = getSideAndAlignFromPlacement(placement);
+    const { arrow: { x: arrowX, y: arrowY } = {} } = middlewareData;
```

Comment:
```suggestion
    const arrowX = middlewareData.arrow?.x;
    const arrowY = middlewareData.arrow?.y;
```

Is this equivalent? I found the nested destructure trickier to follow at first glance.